content/blog/2019-12-02-equivalence-modulo-inputs/index.md

```diff
@@ -0,0 +1,400 @@
+link = "https://www.cs.cornell.edu/~gyauney"
++++
+
+
+Imagine you, a clever person who wants her C programs to run faster,
+sat down, thought very hard, and developed a new compiler optimization.
+Say you implement it as a transformation pass in LLVM so that other people
+can take advantage of your cleverness to make their programs run faster as well.
+You run your optimization pass over a few benchmarks and see that it does
+indeed make some programs run faster.
+But a question nags you: how do you know that your optimization is correct?
+That is, how do you know that your optimization doesn't change the semantics
+of the input program?
+
+*Equivalence Modulo Inputs* (henceforth EMI), a testing technique introduced
+by Le & al. in a [PLDI 2014 paper][paper], allows our compiler hacker
+above to test her optimization rigorously without much effort.
+EMI is especially effective at finding *miscompilation* bugs,
+wherein compilers produce wrong code, which is much more pernicious than
+*compiler crashes*, bugs where the compiler terminates abnormally.
+This allows EMI to test the optimization phase of compilers much more rigorously
+than prior work such as [Csmith][], which mostly finds compiler crashes and not
+miscompilations.
+
+[paper]: https://dl.acm.org/citation.cfm?id=2594334
+[Csmith]: https://dl.acm.org/citation.cfm?id=1993532
+
+
+## Background
+
+EMI is a form of [differential testing][], the widely applicable idea that if
+multiple systems presumed to be equivalent produce different outputs on the same
+input, then there is a bug in at least one of the systems.
+
+Csmith also differentially tests compilers by generating random test cases,
+leading to 281 bug reports for GCC and LLVM by the time it was published.
+Whereas Csmith compares output from multiple compilers on the same input
+program, EMI compares output from different programs compiled by the same
+compiler.
+
+Explicitly wanting to avoid Csmith's painstaking approach to restricting random
+program generation with aggressive safety analysis, Le & al. design their
+implementation of EMI around generating equivalent variants of valid seed
+programs, as we will see below. The EMI and Csmith approaches are not
+oppositional; in fact, Le & al. incorporate Csmith into their workflow. The vast
+majority (all but four) of bugs identified by EMI were found with EMI variants
+of random programs generated by Csmith.
+
+[Differential testing]: https://www.cs.swarthmore.edu/~bylvisa1/cs97/f13/Papers/DifferentialTestingForSoftware.pdf
+
+
+## Some definitions
+
+Let us now formally define EMI and show how we can use it as a condition
+to determine whether a compiler is buggy.
+
+> **Equivalence modulo inputs (EMI)**.
+  Given an input set *I* and a program *P*, a program *Q* is an EMI-variant of
+  *P* (read: *P* and *Q* are EMI) relative to *I* if *P* and *Q* have
+  the same denotations ("mean the same thing") over all inputs in *I*.
+  Formally, for all *i* in *I*, `⟦P⟧(i) = ⟦Q⟧(i)`.
```

Comment:
FWIW, you are allowed to use actual LaTeX math in your blog post if you want! (See the README for this repository.)